### Question

Create a data pipeline for this dataset 
https://www.kaggle.com/datasets/sriharshaeedala/airline-delay/data

which should include following
- Extract (load csv)
- Transform (clean data and transform data if needed)
- Load (save the cleaned data in csv file)

Also create visualization of 
- correlation of attributes
- distribution of values in attributes 
- and any 3 other graph using seaborn 

Note: Also generate insight from it and explain it. 

Submission should be done in github repo with ipynb file or google collab link.


---

### Dataset Description

The dataset is structured as a tabular format with rows representing unique combinations of year, month, carrier, and airport. Each row contains information on various metrics, including flight counts, delay counts, cancellation and diversion counts, and delay breakdowns by different factors. The columns provide specific details such as carrier codes and names, airport codes and names, and counts of delays attributed to carrier, weather, NAS, security, and late aircraft arrivals. The structured format ensures that users can easily query, analyze, and visualize the data to derive meaningful insights.

- year: The year of the data.
- month: The month of the data.
- carrier: Carrier code.
- carrier_name: Carrier name.
- airport: Airport code.
- airport_name: Airport name.
- arr_flights: Number of arriving flights.
- arr_del15: Number of flights delayed by 15 minutes or more.
- carrier_ct: Carrier count (delay due to the carrier).
- weather_ct: Weather count (delay due to weather).
- nas_ct: NAS (National Airspace System) count (delay due to the NAS).
- security_ct: Security count (delay due to security).
- late_aircraft_ct: Late aircraft count (delay due to late aircraft arrival).
- arr_cancelled: Number of flights canceled.
- arr_diverted: Number of flights diverted.
- arr_delay: Total arrival delay.
- carrier_delay: Delay attributed to the carrier.
- weather_delay: Delay attributed to weather.
- nas_delay: Delay attributed to the NAS.
- security_delay: Delay attributed to security.
- late_aircraft_delay: Delay attributed to late aircraft arrival.

## STEP 1: EXTRACT

### 1.1 loading the data from kaggle


In [88]:
# importing the necessary libraries for ETL process
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(style="whitegrid")

In [89]:
# boilerplate code from kaggle to load the dataset
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load
file_path = "Airline_Delay_Cause.csv"

# Load the latest version
df = kagglehub.dataset_load(
  KaggleDatasetAdapter.PANDAS,
  "sriharshaeedala/airline-delay",
  file_path
)

In [90]:
# verifying if data is properly loaded or not
df.head()

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
0,2023,8,9E,Endeavor Air Inc.,ABE,"Allentown/Bethlehem/Easton, PA: Lehigh Valley ...",89.0,13.0,2.25,1.60,...,0.0,5.99,2.0,1.0,1375.0,71.0,761.0,118.0,0.0,425.0
1,2023,8,9E,Endeavor Air Inc.,ABY,"Albany, GA: Southwest Georgia Regional",62.0,10.0,1.97,0.04,...,0.0,7.42,0.0,1.0,799.0,218.0,1.0,62.0,0.0,518.0
2,2023,8,9E,Endeavor Air Inc.,AEX,"Alexandria, LA: Alexandria International",62.0,10.0,2.73,1.18,...,0.0,4.28,1.0,0.0,766.0,56.0,188.0,78.0,0.0,444.0
3,2023,8,9E,Endeavor Air Inc.,AGS,"Augusta, GA: Augusta Regional at Bush Field",66.0,12.0,3.69,2.27,...,0.0,1.57,1.0,1.0,1397.0,471.0,320.0,388.0,0.0,218.0
4,2023,8,9E,Endeavor Air Inc.,ALB,"Albany, NY: Albany International",92.0,22.0,7.76,0.00,...,0.0,11.28,2.0,0.0,1530.0,628.0,0.0,134.0,0.0,768.0


In [91]:
df.tail()

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
171661,2013,8,YV,Mesa Airlines Inc.,TRI,"Bristol/Johnson City/Kingsport, TN: Tri Cities",7.0,1.0,0.78,0.0,...,0.0,0.0,0.0,0.0,23.0,18.0,0.0,5.0,0.0,0.0
171662,2013,8,YV,Mesa Airlines Inc.,YUM,"Yuma, AZ: Yuma MCAS/Yuma International",2.0,1.0,0.00,0.0,...,0.0,1.0,0.0,0.0,17.0,0.0,0.0,0.0,0.0,17.0
171663,2013,8,YV,Mesa Airlines Inc.,SLC,"Salt Lake City, UT: Salt Lake City International",2.0,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
171664,2013,8,DL,Delta Air Lines Inc.,SBN,"South Bend, IN: South Bend International",1.0,1.0,1.00,0.0,...,0.0,0.0,0.0,0.0,18.0,18.0,0.0,0.0,0.0,0.0
171665,2013,8,F9,Frontier Airlines Inc.,MTJ,"Montrose/Delta, CO: Montrose Regional",1.0,0.0,0.00,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### 1.2 performing data validation

In [92]:
# total rows
total_data = len(df)  
total_data

171666

In [93]:
# getting the statistics of numerical columns
display(df.describe().T)

,count,mean,std,min,25%,50%,75%,max
year,171666.0,2018.551361,2.890006,2013.0,2016.00,2019.00,2021.00,2023.00
month,171666.0,6.493633,3.440908,1.0,4.00,7.00,9.00,12.00
arr_flights,171426.0,362.528467,992.894662,1.0,50.00,100.00,250.00,21977.00
arr_del15,171223.0,66.434387,179.540694,0.0,6.00,17.00,47.00,4176.00
carrier_ct,171426.0,20.796615,50.315176,0.0,2.16,6.40,17.26,1293.91
weather_ct,171426.0,2.250347,7.314252,0.0,0.00,0.40,1.86,266.42
nas_ct,171426.0,19.381147,61.675244,0.0,1.00,3.91,11.71,1884.42
security_ct,171426.0,0.157096,0.717405,0.0,0.00,0.00,0.00,58.69
late_aircraft_ct,171426.0,23.770554,72.393477,0.0,1.23,5.00,15.26,2069.07
arr_cancelled,171426.0,7.530497,43.654880,0.0,0.00,1.00,4.00,4951.00


In [94]:
print(df.dtypes)

year                     int64
month                    int64
carrier                 object
carrier_name            object
airport                 object
airport_name            object
arr_flights            float64
arr_del15              float64
carrier_ct             float64
weather_ct             float64
nas_ct                 float64
security_ct            float64
late_aircraft_ct       float64
arr_cancelled          float64
arr_diverted           float64
arr_delay              float64
carrier_delay          float64
weather_delay          float64
nas_delay              float64
security_delay         float64
late_aircraft_delay    float64
dtype: object


In [95]:
# checking if delay minutes and number of delays are same or not and
for i in range(20):
    index = np.random.randint(0, total_data)
    count, cc,wc,nc,sc,lac = df.loc[index, ['arr_del15', 'carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct']]
    mins, cm,wm,nm,sm,lm = df.loc[index, ['arr_delay', 'carrier_delay', 'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay']]
    if mins != round(cm+wm+nm+sm+lm):
        print(f'{index}: Minutes NOT matched')
        print(f'T:{mins} : {cm}, {wm}, {nm}, {sm}, {lm} : {cm+wm+nm+sm+lm}')
    if count != round(cc+wc+nc+sc+lac):
         print(f'{index}: Count NOT matched')
         print(f'T:{count} : {cc}, {wc}, {nc}, {sc}, {lac}: {cc+wc+nc+sc+lac}')
    

In [96]:
# total percentage of data that is null
null_values = df.isna().sum()
print('columns \t\t\t nullvalues%')
for i in range(len(null_values)):
    print(f'{null_values.index[i]} \t\t\t {round(null_values[i]/total_data*100, 3)}%')

columns 			 nullvalues%
year 			 0.0%
month 			 0.0%
carrier 			 0.0%
carrier_name 			 0.0%
airport 			 0.0%
airport_name 			 0.0%
arr_flights 			 0.14%
arr_del15 			 0.258%
carrier_ct 			 0.14%
weather_ct 			 0.14%
nas_ct 			 0.14%
security_ct 			 0.14%
late_aircraft_ct 			 0.14%
arr_cancelled 			 0.14%
arr_diverted 			 0.14%
arr_delay 			 0.14%
carrier_delay 			 0.14%
weather_delay 			 0.14%
nas_delay 			 0.14%
security_delay 			 0.14%
late_aircraft_delay 			 0.14%


/var/folders/lq/wz2w0wfj3gn14q_q15mhm9sc0000gn/T/ipykernel_58462/2139138028.py:5: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f'{null_values.index[i]} \t\t\t {round(null_values[i]/total_data*100, 3)}%')


In [97]:
# looking at random set of data
df.iloc[135:175,:]

,year,month,carrier,carrier_name,airport,airport_name,arr_flights,arr_del15,carrier_ct,weather_ct,...,security_ct,late_aircraft_ct,arr_cancelled,arr_diverted,arr_delay,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay
135,2023,8,AA,American Airlines Inc.,FLL,"Fort Lauderdale, FL: Fort Lauderdale-Hollywood...",447.0,129.0,40.22,3.59,...,0.25,57.76,5.0,3.0,11667.0,3675.0,514.0,1169.0,9.0,6300.0
136,2023,8,AA,American Airlines Inc.,FSD,"Sioux Falls, SD: Joe Foss Field",62.0,25.0,9.74,0.00,...,0.00,10.11,0.0,0.0,2900.0,800.0,0.0,344.0,0.0,1756.0
137,2023,8,AA,American Airlines Inc.,GEG,"Spokane, WA: Spokane International",93.0,25.0,10.35,0.00,...,0.00,10.84,1.0,0.0,3519.0,714.0,0.0,143.0,0.0,2662.0
138,2023,8,AA,American Airlines Inc.,GRR,"Grand Rapids, MI: Gerald R. Ford International",166.0,56.0,17.95,1.69,...,0.00,31.22,2.0,2.0,8476.0,2361.0,76.0,263.0,0.0,5776.0
139,2023,8,AA,American Airlines Inc.,GSO,"Greensboro/High Point, NC: Piedmont Triad Inte...",122.0,33.0,11.03,1.49,...,0.05,16.20,1.0,0.0,3809.0,496.0,169.0,154.0,3.0,2987.0
140,2023,8,AA,American Airlines Inc.,GSP,"Greer, SC: Greenville-Spartanburg International",123.0,44.0,10.40,4.01,...,0.11,26.85,0.0,0.0,3849.0,753.0,324.0,122.0,27.0,2623.0
141,2023,8,AA,American Airlines Inc.,HNL,"Honolulu, HI: Daniel K Inouye International",186.0,41.0,14.52,0.00,...,0.00,11.55,0.0,2.0,2878.0,959.0,0.0,370.0,0.0,1549.0
142,2023,8,AA,American Airlines Inc.,HSV,"Huntsville, AL: Huntsville International-Carl ...",9.0,5.0,1.28,0.00,...,0.00,3.70,0.0,0.0,581.0,125.0,0.0,3.0,0.0,453.0
143,2023,8,AA,American Airlines Inc.,IAD,"Washington, DC: Washington Dulles International",76.0,12.0,4.85,0.75,...,0.00,5.73,0.0,0.0,834.0,338.0,49.0,23.0,0.0,424.0
144,2023,8,AA,American Airlines Inc.,IAH,"Houston, TX: George Bush Intercontinental/Houston",683.0,180.0,75.08,8.10,...,0.00,77.69,8.0,1.0,17333.0,7728.0,838.0,1105.0,0.0,7662.0


In [98]:
# checking if unique or not
pk_columns = ['year', 'month', 'carrier', 'airport']
duplicate_count = df.duplicated(subset=pk_columns).sum()
duplicate_count

np.int64(0)

In [101]:
delay_ct_cols = ['carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct']
delay_mins_cols = ['carrier_delay', 'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay']

zero_delay_ct_rows = df[(df[delay_ct_cols] == 0).all(axis=1)]
zero_delay_mins_rows = df[(df[delay_mins_cols] == 0).all(axis=1)]
zero_total_delay_ct = df[(df['arr_delay'] == 0)]
zero_total_delay_mins = df[(df['arr_del15'] == 0)]

print(f"Rows where all delay causes count is 0: {len(zero_delay_ct_rows)} ")
print(f"Total 0 count: {len(zero_total_delay_ct)}")
print("---")
print(f"Rows where all delay mins is 0: {len(zero_delay_mins_rows)} ")
print(f"Total 0 mins: {len(zero_total_delay_mins)}")


Rows where all delay causes count is 0: 6788 
Total 0 count: 6788
---
Rows where all delay mins is 0: 6788 
Total 0 mins: 6585


In [ ]:
# so basically causes count is fine, however the total time is kindof illogical around 200 data points


In [103]:
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns
all_zero_numeric = df[
    (df[numeric_cols] == 0).all(axis=1)
]

print("Rows where all numeric values are zero:", len(all_zero_numeric))

Rows where all numeric values are zero: 0


In [104]:
invalid_delay_rows = df[df['arr_del15'] > df['arr_flights']]
print("Invalid delay rows:", len(invalid_delay_rows))

Invalid delay rows: 0


### From initial exploration, following things are deduced:

1. Total rows: 171,666
2. All columns are in the respective datatypes i.e. numerical columns in int and float, string columns in object
3. For non-null values, the total delay count and total delay minutes is exactly the sum of different causes of counts of delay and minutes of delay
4. Around 0.14% of most columns is null, however max is of arr_del15 with 0.258%
5. Primary key for this dataset: (year, month, carrier, airport)
6. NO duplicate values
7. 6788 instances had no delay. However, 6585 instances are sure to have no delays but the random 200-ish datapoints seem illogical
8. almost all data integrity checks show that the data is good.

---

## STEP 2: TRANSFORM